src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,358 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Code to migrate from a non-Timely-friendly remap shard schema to a
+//! Timely-friendly one.
+use std::sync::Arc;
+
+use anyhow::Context;
+use differential_dataflow::consolidation;
+use differential_dataflow::lattice::Lattice;
+use itertools::Itertools;
+use mz_ore::now::EpochMillis;
+use mz_persist_types::Codec64;
+use timely::order::{PartialOrder, TotalOrder};
+use timely::progress::frontier::{Antichain, MutableAntichain};
+use timely::progress::Timestamp;
+use tokio::sync::Mutex;
+use tracing::info;
+
+use mz_expr::PartitionId;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::ReadHandle;
+use mz_persist_client::{PersistClient, ShardId, Upper};
+use mz_repr::{Datum, Diff, GlobalId, TimestampManipulation};
+use mz_timely_util::order::Partitioned;
+
+use crate::types::sources::data::SourceData;
+use crate::types::sources::MzOffset;
+use crate::util::antichain::MutableOffsetAntichain;
+use crate::util::remap_handle::RemapHandle;
+
+use super::CollectionMetadata;
+
+/// A struct capable of performing remap shard migrations. See comment on
+/// [`RemapHandleMigrator::migrate`].
+pub struct RemapHandleMigrator<
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
+    persist_client: PersistClient,
+    current_shard_read_handle: ReadHandle<SourceData, (), T, Diff>,
+    as_of: Antichain<T>,
+    write_shard: ShardId,
+    partitioned_source: bool,
+}
+
+impl<T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation>
+    RemapHandleMigrator<T>
+{
+    /// Prior to this commit, remap shards contained data of schema (Int32?,
+    /// Uint64). Int32 was present only for Kafka sources where it represented
+    /// the Kafka partition.
+    ///
+    /// This schema introduced an impedance mismatch between the remap data and
+    /// native Timely types; tl;dr is that an empty remap shard, if taken
+    /// directly to Timely time types, produces an empty Antichain (i.e. the
+    /// shard is closed from further modification), when in reality we wanted
+    /// it to express the minimum Antichain (i.e. any future input is valid).
+    ///
+    /// TO resolve this, we designed a migration to instead use the schema
+    /// (Range<Int32>?, Uint64). Range<Int32> still represents Kafka
```

Comment:
That's going to have the unfortunate side effect of making the remap collection data challenging to use because you can't use currently generic numeric input (e.g. `1` which is `Int32`) with `Range<UInt32>`. For example, the contains operator requires that the element type you're searching for be implicitly castable to the range's element type, and `Int32` cannot be implicitly cast to `UInt32`.

We work around this in the case of binary operations on `Int32` and `UInt32` by casting both to `Int64`, however, that code path doesn't currently work for polymorphic arguments, e.g.

```
select list_cat(LIST[1,2,3]::uint4 list, 4);
ERROR:  function list_cat(uint4 list, integer) does not exist
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

We could make this work similarly to how it does for binary operations, but would require casting the range to `Range<Int64>` and the `Int32` to `Int64`.

If the apprehension is moreso that negative numbers could creep in, it would be more ergonomic with the current type system to just make that an assertion instead.